File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
         if(root == NULL)
        return 0;
        long long ans = 0;
        queue<pair<TreeNode*, long long>> q;
        q.push({root,0});
        
        while(!q.empty()){
            int size = q.size();
            int mini = q.front().second;
            long long first, last;
            for(int i=0;i<size;i++){
                TreeNode* r = q.front().first;
                long long id = q.front().second - mini;
                q.pop();

                if(i == 0)first = id;
                if(i == size-1)last = id;

                if(r->left != NULL)q.push({r->left , id*2+1});
                if(r->right != NULL)q.push({r->right , id*2+2});
            }
            ans = max(ans,last-first+1);
        }
        
        return (int)ans;
    }
};

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.java
class Pair{
    TreeNode first;
    long second;
    Pair(TreeNode first, long second){
        this.first  = first;
        this.second = second;
    }
}
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        if(root == null)return 0;
        long ans = 0;
        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(root,0));

        while(!q.isEmpty()){
            int size = q.size();
            long min  = q.peek().second;

            long first=0,last=0;
            for(int i=0;i<size;i++){
                TreeNode r = q.peek().first;
                long id = q.peek().second;
                q.poll();

                if(i == 0)first = id;
                if(i == size-1)last = id;

                if(r.left != null)q.offer(new Pair(r.left,id*2+1));
                if(r.right != null)q.offer(new Pair(r.right,id*2+2));
            }
            ans = Math.max(last-first+1,ans);
        }
        return (int)ans;
    }
}